### PR TITLE
Implement Email Queue

### PR DESCRIPTION
**Pull request by rymaju**
Worked with Joe on this!

Branched off #30, dependent on that merging first.

# Summary
Currently if used the EmailService we could get throttled by Amazon SES leading to dropped emails. This PR implements a different system that ~instead adds emails to a working queue, which gets processed by a cron job. This cron job sends the first 14 emails from the queue every second. This should prevent throttling from Amazon.~

...uses the great `bottleneck` library to ensure we stay within the rate limit.

However, this does not protect against daily limits of 50,000 emails sent per day. If we hit this limit, we would be continually erroring, failing to send all 14 emails per second, ~which all would be requeued to be sent again in the next second.~

Emails that fail are simply dropped and logged. Since we cant really afford to be dropping emails, a more robust system might be marking users who we have successfully sent emails to.

I don't think we need that system though, in the worst case we need to sort through a ton of error logs and set up a batch processing system to handle retrying failed attempts.

## On Testing

The only new logic for this service is the rate limiting, I couldn't think of any other business logic that could be tested. I think this is sufficient. Although, we should definitely have some kind of load/performance testing in the future.

### CODE REVIEW

**Comment by rymaju** [comment]
In retrospect, is this overkill? Will we ever get >14 emails sent per second? I have to think about this... 🤔

**Comment by rymaju** [comment]
https://docs.aws.amazon.com/ses/latest/DeveloperGuide/manage-sending-quotas.html

**Comment by rymaju** [comment]
https://docs.aws.amazon.com/ses/latest/dg/manage-sending-quotas-request-increase.html

I think we do need this 14 per second system but it looks like AWS will automagically increase the limit as we approach it

File: src/util/email/email.service.ts [hunk 1]
```diff
@@ -1,24 +1,90 @@
...
+  /**
+   * Emails are placed in a queue and only processed `EMAILS_SENT_PER_SECOND` emails per second
+   * in order to avoid rate-limiting.
+   *
+   * Immutable.js Lists implement an efficient Queue data structure.
+   */
+  private emailsToSend: List<SendEmailTask>; // Generic Data (Parameterized Types, Dependent Types)
+
+  constructor(private amazonSESWrapper: AmazonSESWrapper) {
+    this.emailsToSend = List(); // Array(150_000).fill(sendEmailTaskExample)
```
Comment (onb6) [inline src/util/email/email.service.ts:20]: what does this comment mean?

File: src/util/email/email.service.ts [hunk 2]
```diff
@@ -1,24 +1,90 @@
...
+   * @param bodyHTML the HTML body of the email
+   */
+  public queueEmail(recipientEmail: string, subject: string, bodyHTML: string): void {
+    this.emailsToSend = this.emailsToSend.push({ recipientEmail, subject, bodyHTML });
+  }
+
+  /**
+   * Runs every second.
+   * Processes the first `this.EMAILS_SENT_PER_SECOND` emails in the queue in order to avoid sending too many emails at once.
+   * If the queue is empty, does nothing.
```
Comment (onb6) [inline src/util/email/email.service.ts:36]: coming from a place of knowing next to nothing about cron: this running every second won't slow down the site, right?

File: src/util/email/email.service.ts [hunk 3]
```diff
@@ -1,24 +1,90 @@
...
    */
-  public async sendEmails(
-    recipientEmails: string[],
+  private async sendEmail(
+    recipientEmail: string,
     subject: string,
     bodyHTML: string,
   ): Promise<unknown> {
-    return this.amazonSESWrapper.sendEmails(recipientEmails, subject, bodyHTML);
+    return this.amazonSESWrapper.sendEmails([recipientEmail], subject, bodyHTML);
```
Comment (onb6) [inline src/util/email/email.service.ts:82]: clarifying question: im assuming the sendEmails() method from SES requires a list here?

File: src/util/email/email.service.ts [hunk 4]
```diff
@@ -1,24 +1,90 @@
...
+    recipientEmail,
+    subject,
+    bodyHTML,
+  }: SendEmailTask): Promise<void> {
+    try {
+      const result = await this.sendEmail(recipientEmail, subject, bodyHTML);
+      this.logger.log(`Sent email to ${recipientEmail} with result: ${result}`);
+    } catch (error) {
+      this.logger.error(`Failed to send email to ${recipientEmail}`, error);
+      this.queueEmail(recipientEmail, subject, bodyHTML);
```
Comment (onb6) [inline src/util/email/email.service.ts:64]: i like this

**Review by onb6 — APPROVED** [review]
left some probably dumb questions, this is great though! glad we were able to solve this with so little code :)

**Comment by onb6** [comment]
> 

maybe, but better safe than sorry!

**Comment by onb6** [comment]
is there a way to test this?

**Comment by rymaju** [comment]
No these are great questions, I don't feel comfortable merging this until I have a better idea of how to test this and how this will affect performance. Especially if these requests hang, it could eat up a lot of space in the event loop. Maybe we should timeout these promises?

Testing is also a pain, not sure if theres a great way to test this but until I can think of a way we should also wait on merging this in.

File: src/util/email/email.service.ts (hunk 2 above)
Comment (rymaju) [reply src/util/email/email.service.ts:36]: A cron job running every second would be scary if the operation was very CPU intensive, but this should just be sending (at most) 14 HTTP requests async. Our server should be able to handle that?

File: src/util/email/email.service.ts (hunk 1 above)
Comment (rymaju) [reply src/util/email/email.service.ts:20]: This was to test that this even worked at all, we should move this into a real test 😓

**Pull request merged**